src/main/java/org/openstreetmap/atlas/tags/filters/matcher/parsing/Parser.java

```diff
@@ -0,0 +1,464 @@
+import org.openstreetmap.atlas.tags.filters.matcher.parsing.tree.UnaryOperator;
+import org.openstreetmap.atlas.tags.filters.matcher.parsing.tree.XorOperator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class can transform a sequence of {@link Token}s into a syntactically valid abstract syntax
+ * tree (AST) that can then be checked by the {@link SemanticChecker}. Finally, a
+ * {@link TaggableMatcher} may walk this AST to determine if a {@link Taggable}'s tag map
+ * corresponds to the matcher. This {@link Parser} implements an LL(1) {@link TaggableMatcher}
+ * expression grammar using recursive descent. The grammar can be found below in comment form.
+ *
+ * @author lcram
+ */
+public class Parser
+{
+    /*
+     * The Grammar. Operator precedence is handled in the standard way. '=' and '!=' are treated as
+     * extremely "sticky" (i.e. high precedence) operators. The grammar is not capable of detecting
+     * "nested" '=' and '!=' operators (e.g. foo=(bar=baz)), which are syntactically valid but
+     * semantically invalid. Syntax trees containing nested equality operators must be dealt with at
+     * a later stage.
+     */
+
+    // OR -> XOR OR'
+    // OR' -> | XOR OR'
+    // OR' -> ''
+    // XOR -> AND XOR'
+    // XOR' -> | AND XOR'
+    // XOR' -> ''
+    // AND -> EQ AND'
+    // AND' -> & EQ AND'
+    // AND' -> ''
+    // EQ -> VALUE EQ'
+    // EQ' -> = VALUE EQ'
+    // EQ' -> != VALUE EQ'
+    // EQ' -> ''
+    // VALUE -> ( OR )
+    // VALUE -> ! VALUE
+    // VALUE -> literal
+    // VALUE -> /regex/
+
+    /**
+     * @author lcram
+     */
+    private static class TokenBuffer
+    {
+        private final List<Token> tokens;
+        private final String inputLine;
+        private int position;
+
+        TokenBuffer(final List<Token> tokens, final String inputLine)
+        {
+            this.tokens = tokens;
+            this.inputLine = inputLine;
+            this.position = 0;
+        }
+
+        void nextToken()
+        {
+            if (this.position < this.tokens.size())
+            {
+                this.position++;
+            }
+        }
+
+        Token peek()
+        {
+            if (this.position >= this.tokens.size())
+            {
+                return new Token(Token.TokenType.EOF, null, this.inputLine.length());
+            }
+            return this.tokens.get(this.position);
+        }
+    }
+
+    private static final Logger logger = LoggerFactory.getLogger(Parser.class);
+
+    private final TokenBuffer tokenBuffer;
+    private final String inputLine;
+
+    public Parser(final List<Token> tokens, final String inputLine)
+    {
+        this.tokenBuffer = new TokenBuffer(tokens, inputLine);
+        this.inputLine = inputLine;
+    }
+
+    public ASTNode parse()
+    {
+        BinaryOperator.clearIdentifierCounter();
+        UnaryOperator.clearIdentifierCounter();
+        Operand.clearIdentifierCounter();
+        return or();
```

Comment:
Aaah you start with or I see

Reply:
Yes! If you check the grammar at the top of the class you can see that the OR transition is the first one. So we start recursive descent with that rule.